File: morning_exercises/day19.cpp
#include <climits>
#include <iostream>
#include <memory>
#include <vector>

class BinaryNode {
    public:
        int val = 0;
        std::shared_ptr<BinaryNode> left = nullptr;
        std::shared_ptr<BinaryNode> right = nullptr;
    
    BinaryNode(int val): val(val) {}

    static int maxPathSumHelper(std::shared_ptr<BinaryNode> root, std::vector<int>& globalMax) {
        if (root==nullptr) return 0;

        int bestLeft = std::max(0, maxPathSumHelper(root->left, globalMax));
        int bestRight = std::max(0, maxPathSumHelper(root->right, globalMax));
        globalMax[0]=std::max(globalMax[0], root->val+bestLeft+bestRight);

        return root->val+std::max(bestLeft,bestRight);
    }

    static int maxPathSum(std::shared_ptr<BinaryNode> root) {
        std::vector<int> globalMax = {INT_MIN};

        maxPathSumHelper(root, globalMax);

        std::cout << maxPathSumHelper(root, globalMax) << "\n";

        return globalMax[0];
    }
};

int main() {
    auto root = std::make_shared<BinaryNode>(11);
    root->left = std::make_shared<BinaryNode>(2);
    root->right = std::make_shared<BinaryNode>(3);
    root->left->left = std::make_shared<BinaryNode>(4);
    root->left->right = std::make_shared<BinaryNode>(5);

    std::cout << BinaryNode::maxPathSum(root) << "\n";
}

File: morning_exercises/day19.py
from __future__ import annotations

class BinaryNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None

    @staticmethod
    def maxPathSumHelper(root: BinaryNode | None, globalMax: list)->int:
        if root is None: return 0

        leftBest = max(0, BinaryNode.maxPathSumHelper(root.left, globalMax))
        rightBest = max(0, BinaryNode.maxPathSumHelper(root.right, globalMax))

        globalMax[0] = max(globalMax[0], root.val + leftBest + rightBest)

        return root.val + max(leftBest, rightBest)
    
    @staticmethod
    def maxPathSum(root: BinaryNode | None):
        globalMax = [float('-inf')]

        BinaryNode.maxPathSumHelper(root, globalMax)

        return globalMax[0]
    
root = BinaryNode(11)
root.left = BinaryNode(2)
root.right = BinaryNode(3)
root.left.left = BinaryNode(4)
root.left.right = BinaryNode(5)

print(BinaryNode.maxPathSum(root))
